pkg/grpc/server.go

```diff
@@ -0,0 +1,180 @@
+// Copyright 2017 Envoyproxy Authors
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+// Package server provides an implementation of a streaming xDS server
+package server
+
+import (
+	"strconv"
+	"sync/atomic"
+
+	"github.com/envoyproxy/go-control-plane/api"
+	"github.com/envoyproxy/go-control-plane/pkg/cache"
+	context "golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// Server is a collection of handlers for streaming discovery requests
+type Server interface {
+	// Register the handlers in a gRPC server
+	Register(*grpc.Server)
+}
+
+func NewServer(config cache.ConfigCache) Server {
+	return &server{config: config}
+}
+
+type server struct {
+	config cache.ConfigCache
+}
+
+func (s *server) Register(srv *grpc.Server) {
+	api.RegisterAggregatedDiscoveryServiceServer(srv, s)
+	api.RegisterEndpointDiscoveryServiceServer(srv, s)
+	api.RegisterClusterDiscoveryServiceServer(srv, s)
+	api.RegisterRouteDiscoveryServiceServer(srv, s)
+	api.RegisterListenerDiscoveryServiceServer(srv, s)
+}
+
+type stream interface {
+	Send(*api.DiscoveryResponse) error
+	Recv() (*api.DiscoveryRequest, error)
+}
+
+func (s *server) handler(stream stream, selector cache.ResourceSelector) error {
+	// a channel for receiving incoming requests
+	reqCh := make(chan *api.DiscoveryRequest, 0)
+	reqStop := int32(0)
+	go func() {
+		for {
+			req, err := stream.Recv()
+			if atomic.LoadInt32(&reqStop) != 0 {
+				return
+			}
+			if err != nil {
+				close(reqCh)
+				return
+			}
+			reqCh <- req
+		}
+	}()
+	// prevents writing to a closed channel if send failed on blocked recv
+	defer func() {
+		atomic.StoreInt32(&reqStop, 1)
+	}()
+
+	// the node issuing the last request (assumed to be constant for the session)
+	var node *api.Node
+
+	// unique nonce for req-resp pairs; the server ignores stale nonces
+	var nonce int64
+
+	// last successfully applied version as set in the requests
+	var version string
+
+	for {
+		config := cache.Promise{}
+		defer func() {
+			if config.Stop != nil {
+				config.Stop()
+			}
+		}()
```

Comment:
Should this be lifted out of the loop body? We really only need to call `Stop` on the last config request in flight, since all previous lookups either finished successfully (no need to call `Stop`), or we updated the request and called `Stop` in the loop already, right?

```go
var config cache.Promise
defer func() {
	if config.Stop != nil {
		config.Stop()
	}
}()

for {
	// make a cache request
	if node != nil {
		config = s.config.Listen(node, selector, version)
	}

	// ...
}

```


(BTW @kyessenov I'm shifting over to work on Pilot full time, looking at what needs to change in Pilot config gen world in light of xDS v2.)